crates/shared/src/caching_balance_fetcher.rs

```diff
@@ -0,0 +1,146 @@
+use crate::{
+    account_balances::{BalanceFetching, Query, TransferSimulationError},
+    current_block::{into_stream, CurrentBlockStream},
+};
+use anyhow::Result;
+use futures::StreamExt;
+use itertools::Itertools;
+use model::order::SellTokenSource;
+use primitive_types::{H160, U256};
+use std::{
+    collections::HashMap,
+    sync::{Arc, RwLock},
+};
+use tracing::Instrument;
+
+#[derive(Default)]
+struct BalanceCache {
+    /// Block number the cached data was collected from.
+    block: u64,
+    data: HashMap<Query, U256>,
```

Comment:
Can't we use the `RequestSharing` to synchronize updates and requests?

Reply:
Yeah, I guess I could just put that in the `Web3BalanceFetcher` as a follow up PR.

Reply:
Perfect! Great idea.